src/plugins/share/common/url_service/url_service.ts

```diff
@@ -7,19 +7,25 @@
  */
 
 import { LocatorClient, LocatorClientDependencies } from './locators';
+import { IShortUrlClientFactory } from './short_urls';
 
-export type UrlServiceDependencies = LocatorClientDependencies;
+export interface UrlServiceDependencies<D = unknown> extends LocatorClientDependencies {
+  shortUrls: IShortUrlClientFactory<D>;
+}
 
 /**
  * Common URL Service client interface for server-side and client-side.
  */
-export class UrlService {
+export class UrlService<D = unknown> {
```

Comment:
why do we need generic here ? seems we know deps.shortUrls needs to be there and we also know all deps for LocatorClient ? also having D being unknown by default is a bit weird and not fully type safe.

Reply:
> why do we need generic here ? seems we know deps.shortUrls needs to be there and we also know all deps for LocatorClient ?

It is generic because Short URL service has different dependencies on the server vs on the browser.

> also having D being unknown by default is a bit weird and not fully type safe.

Why do you think it is weird and not safe?